### ISSUE TITLE
`suspicious_arithmetic_impl` flags on size addition in multiplication

### ISSUE THREAD

**Issue by mnemonikr** [rust-lang/rust-clippy#16247]
### Summary

I maintain a library that supports symbolic multiplication on bitvectors. The API exposes how many bits to keep in the output, i.e. `multiply(lhs, rhs, n)`. For the `std::ops::Mul` impl I just use the maximum size needed: `multiply(lhs, rhs, lhs.len() + rhs.len())`. Because the lint sees the `+` as the sole binary operator in `Mul`, it flags it as suspicious.

As has been noted in other issues (#2268, #3215), this lint is hard to get right, but would be great if it didn't flag here.

### Lint Name

suspicious_arithmetic_impl

### Reproducer

I tried this code:

```rust
pub struct BitVec {
    bytes: Vec<u8>,
}

impl BitVec {
    pub fn multiply(self, _rhs: Self, _n: usize) -> Self {
        unimplemented!()
    }
}

impl std::ops::Mul for BitVec {
    type Output = Self;

    fn mul(self, rhs: Self) -> Self::Output {
        // False positive due to +
        let n = self.bytes.len() + rhs.bytes.len();
        self.multiply(rhs, n)
    }
}
```

I saw this happen:

```
warning: suspicious use of `+` in `Mul` impl
   --> test.rs:16:34
    |
16 |         let n = self.bytes.len() + rhs.bytes.len();
    |                                  ^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.91.0/index.html#suspicious_arithmetic_impl
    = note: `#[warn(clippy::suspicious_arithmetic_impl)]` on by default
```

I expected to see this happen: Lint not to fire.

### Version

```text
rustc 1.91.1 (ed61e7d7e 2025-11-07)
binary: rustc
commit-hash: ed61e7d7e242494fb7057f2657300d9e77bb4fcb
commit-date: 2025-11-07
host: x86_64-unknown-linux-gnu
release: 1.91.1
LLVM version: 21.1.2
```

**Comment by samueltardieu** [comment]
This looks like a perfect opportunity to apply `#[expect(clippy::suspicious_arithmetic_impl, reason = "addition of length is required")]` on your method (or just `#[allow]` if you prefer).

**Comment by mnemonikr** [comment]
Haha yes, first thing I did was suppress the lint at the flagged location. Though I now understand the heuristic it applied in order to flag it, I must admit I was bewildered when this was flagged. Clippy is usually so on point.
